sumpy/recurrence.py

```diff
@@ -0,0 +1,100 @@
+__copyright__ = """
+Copyright (C) 2024 Andreas Kloeckner
+Copyright (C) 2024 Hirish Chandrasekaran
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from collections import namedtuple
+from pyrsistent import pmap
+from pytools import memoize
+from sumpy.tools import add_mi
+from itertools import accumulate
+import sumpy.symbolic as sym
+import logging
+from typing import List
+import sympy as sp
+from sumpy.expansion.diff_op import LinearPDESystemOperator
+from pytools.obj_array import make_obj_array
+
+
+__doc__ = """
+.. autoclass:: Recurrence
+
+.. autofunction:: make_sympy_vec
+"""
+
+def make_sympy_vec(name, n):
+    return make_obj_array([sp.Symbol(f"{name}{i}") for i in range(n)])
+
+class Recurrence:
```

Comment:
Is there a good reason this should be a class? (vs. just a function)

Reply:
I think so. Right now the code is structured so that you have a sequence of functions where the output of one function goes into the input of another function. There are intermediate representations that I am interested in like the ODE form of the PDE. If I create a recurrence class then I can store all the intermediate representations as attributes and get them immediately if I am interested in them.